boxsdk/object/upload_session.py

```diff
@@ -0,0 +1,167 @@
+# coding: utf-8
+from __future__ import unicode_literals, absolute_import
+
+import base64
+import hashlib
+import json
+
+from .base_object import BaseObject
+from ..config import API
+from ..pagination.chunked_upload_part_limit_offset_based_object_collection import ChunkedUploadPartLimitOffsetBasedObjectCollection
+
+
+class UploadSession(BaseObject):
+    _item_type = 'upload_session'
+    _parent_item_type = 'file'
+
+    def get_url(self, *args):
+        """
+        Base class override. Endpoint is a little different - it's /files/upload_sessions.
+
+        :rtype:
+            `unicode`
+        """
+        return self.session.get_url(
+            '{0}s/{1}s'.format(self._parent_item_type, self._item_type),
+            self._object_id,
+            *args
+        ).replace(API.BASE_API_URL, API.UPLOAD_URL)
+
+    def get_parts(self, limit=None, offset=None, fields=None):
+        """
+        Get a list of parts uploaded so far.
+
+        :param limit:
+            The maximum number of items to return per page. If not specified, then will use the server-side default.
+        :type limit:
+            `int` or None
+        :param offset:
+            The index at which to start returning items.
+        :type offset:
+            `int` or None
+        :param fields:
+            Fields to include on the returned items.
+        :type fields:
+            `Iterable` of `unicode`
+        :returns:
+            Returns a `list` of parts uploaded so far.
+        :rtype:
+            `list` of `dict`
+        """
+        return ChunkedUploadPartLimitOffsetBasedObjectCollection(
+            session=self.session,
+            url=self.get_url('parts'),
+            limit=limit,
+            fields=fields,
+            offset=offset,
+            return_full_pages=False,
+        )
+
+    def upload_part(self, part_bytes, offset, total_size, part_content_sha1=None):
+        """
+        Upload a part of a file.
+
+        :param part_bytes:
+            Part bytes
+        :type part_bytes:
+            `bytes`
```

Comment:
This is a pretty unfriendly interface, since it doesn't allow streaming and requires the entire part content to stay in memory until it's completely uploaded.

I'd suggest keeping the interface as a content stream, and perhaps including a convenience method that takes `bytes`. That method could wrap them in a `BytesIO` and call the method that takes a content stream.